src/renderers/shared/fiber/ReactFiberBeginWork.js

```diff
@@ -277,6 +277,11 @@ module.exports = function<T, P, I, TI, PI, C, CX, PL>(
     markRef(current, workInProgress);
 
     if (!shouldUpdate) {
+      // Context providers should defer to sCU for rendering
+      if (hasContext) {
+        invalidateContextProvider(workInProgress, false);
```

Comment:
That means that we have to push/pop every functional component though, right? That seems like we're severely punishing perf of functional components for this edge case.